File: sorting/selection_sort/selection_sort.py
# find the smallest integer
def find_smallest(arr): 
    smallest = arr[0]
    smallest_index = 0
    for i in range (1, len(arr)):
        if(arr[i]< smallest):
            smallest = arr[i]
            smallest_index = i
    return smallest_index

# sort by finding smallest integer and insert the same into new array
def selection_sort(arr):
    newArr = []
    copiedArray = list(arr)
    for i in range(len(copiedArray)):
        smallest = find_smallest(copiedArray)
        newArr.append(copiedArray.pop(smallest))
        print("new array after process")
        print(newArr)
    return newArr

print(selection_sort([5,7,1,2,8,3,9]))

File: sorting/selection_sort/selection_sort.go
package main

import (
    "errors"
    "fmt"
)

func main() {
    fmt.Println("Selection sort start")
    arr := []int{5, 8, 1, 4, 2, 9, 1}
    sortedArr := selectionSort(arr)
    fmt.Println(sortedArr) // Print the sorted array
}

func selectionSort(arr []int) []int {
    newArr := make([]int, 0, len(arr)) // Pre-allocate newArr for efficiency
    for len(arr) > 0 {
        smallest, err := findSmallest(arr)
        if err != nil {
            panic(err) // Handle the error appropriately
        }
        newArr = append(newArr, arr[smallest])
        arr = append(arr[:smallest], arr[smallest+1:]...)
    }
    return newArr
}

func findSmallest(arr []int) (int, error) {
    if len(arr) == 0 {
        return 0, errors.New("Empty Array")
    }
    smallestIndex := 0
    for i, v := range arr {
        if v < arr[smallestIndex] {
            smallestIndex = i
        }
    }
    return smallestIndex, nil
}
